rmf_task/include/rmf_task/Task.hpp

```diff
@@ -0,0 +1,295 @@
+    rmf_traffic::Time original_finish_estimate_);
+
+  /// The booking information of the request that this Task is carrying out
+  const ConstBookingPtr& booking() const;
+
+  /// The category of this Task.
+  const std::string& category() const;
+
+  /// Details about this Task.
+  const std::string& detail() const;
+
+  /// The original finish estimate of this Task.
+  rmf_traffic::Time original_finish_estimate() const;
+
+  class Implementation;
+private:
+  rmf_utils::impl_ptr<Implementation> _pimpl;
+};
+
+//==============================================================================
+/// An abstract interface for computing the estimate and invariant durations
+/// of this request
+class Task::Model
+{
+public:
+
+  /// Estimate the state of the robot when the task is finished along with
+  /// the time the robot has to wait before commencing the task
+  virtual std::optional<Estimate> estimate_finish(
+    const State& initial_state,
+    const Constraints& task_planning_constraints,
+    const TravelEstimator& travel_estimator) const = 0;
+
+  /// Estimate the invariant component of the task's duration
+  virtual rmf_traffic::Duration invariant_duration() const = 0;
+
+  virtual ~Model() = default;
+};
+
+//==============================================================================
+/// An abstract interface to define the specifics of this task. This
+/// implemented description will differentiate this task from others.
+class Task::Description
+{
+public:
+
+  /// Generate a Model for the task based on the unique traits of this
+  /// description
+  ///
+  /// \param[in] earliest_start_time
+  ///   The earliest time this task should begin execution. This is usually
+  ///   the requested start time for the task.
+  ///
+  /// \param[in] parameters
+  ///   The parameters that describe this AGV
+  virtual ConstModelPtr make_model(
+    rmf_traffic::Time earliest_start_time,
+    const Parameters& parameters) const = 0;
+
+  // Virtual destructor
+  virtual ~Description() = default;
+};
+
+//==============================================================================
+class Task::Active
+{
+public:
+  /// Backup data for the task. The state of the task is represented by a
+  /// string. The meaning and format of the string is up to the Task
+  /// implementation to decide.
+  ///
+  /// Each Backup is tagged with a sequence number. As the Task makes progress,
+  /// it can issue new Backups with higher sequence numbers. Only the Backup
+  /// with the highest sequence number will be kept.
+  using Backup = detail::Backup;
+
+  /// Descriptions of the phases that have been completed
+  virtual const std::vector<Phase::ConstCompletedPtr>&
+  completed_phases() const = 0;
+
+  /// Interface for the phase that is currently active
+  virtual Phase::ConstActivePtr active_phase() const = 0;
+
+  /// Descriptions of the phases that are expected in the future
+  virtual std::vector<Phase::Pending> pending_phases() const = 0;
```

Comment:
I looked it over, and I did intend for this to be `Phase::Pending` because `Phase::Pending` is a concrete class, not an abstract interface.

However, I've decided to change the overall return value to an immutable reference to avoid unnecessary copying: 7d8db73